op_crates/crypto/01_crypto.js

```diff
@@ -1,25 +1,211 @@
+
+  class RegisteredAlgorithmsContainer {
+    #nameIndex;
+    #definitions;
+
+    constructor(definitions) {
+      this.#nameIndex = Object.create(null);
+      this.#definitions = Object.create(null);
+      for (const [name, definition] of Object.entries(definitions)) {
+        this.#nameIndex[name.toUpperCase()] = name;
+        this.#definitions[name] = definition;
+      }
+    }
+
+    getDefinition(name) {
+      const upperCaseName = toUpperCase.call(name);
+      const registeredName = this.#nameIndex[upperCaseName];
+      if (registeredName === undefined) {
+        return undefined;
+      }
+      return this.#definitions[registeredName];
+    }
+
+    normalizeName(providedName) {
+      const upperCaseName = toUpperCase.call(providedName);
+      return this.#nameIndex[upperCaseName];
+    }
+  }
+
+  const supportedAlgorithms = {
+    digest: new RegisteredAlgorithmsContainer({
+      "SHA-1": {},
+      "SHA-256": {},
+      "SHA-384": {},
+      "SHA-512": {},
+    }),
+  };
+
+  const digestByteLengths = {
+    "SHA-1": 20,
+    "SHA-256": 32,
+    "SHA-384": 48,
+    "SHA-512": 64,
+  };
+
+  function normalizeAlgorithm(algorithm, registeredAlgorithms) {
+    let alg;
+    if (typeof algorithm === "string") {
+      alg = { name: algorithm };
+    }
+    if (typeof algorithm === "object" && algorithm !== null) {
+      if (typeof algorithm.name !== "string") {
+        throw new TypeError("Algorithm name is missing or not a string");
+      }
+      const algorithmName = registeredAlgorithms.normalizeName(algorithm.name);
+      if (algorithmName === undefined) {
+        throw new DOMException(
+          "Unrecognized algorithm name",
+          "NotSupportedError",
+        );
+      }
+      const definition = registeredAlgorithms.getDefinition(algorithmName);
+      // TODO: Validate the algorithm data based on the definition
+      alg = { name: algorithmName };
+    } else {
+      throw new TypeError("Argument 1 must be an object or a string");
+    }
+    return alg;
+  }
+
+  const subtle = {
+    async decrypt(algorithm, key, data) {
+      await undefined;
+      throw new Error("Not implemented");
+    },
+    async deriveBits(algorithm, baseKey, length) {
+      await undefined;
+      throw new Error("Not implemented");
+    },
+    async deriveKey(
+      algorithm,
+      baseKey,
+      derivedKeyType,
+      extractable,
+      keyUsages,
+    ) {
+      await undefined;
+      throw new Error("Not implemented");
+    },
+    async digest(algorithm, data) {
+      const alg = normalizeAlgorithm(algorithm, supportedAlgorithms.digest);
+      // TODO: validate that data is indeed a BufferSource
+      // Note: we need to copy the data because the specs require it.
+      // It's unclear whether we are better doing it here or on the Rust side.
+      const input = data.slice(0);
+      const output = new ArrayBuffer(digestByteLengths[alg.name]);
+      await core.jsonOpAsync(
+        "op_subtle_digest",
```

Comment:
I would prefer if all ops in the crypto op crate would start with with `op_crypto`.

Reply:
Shall I also rename the one for `getRandomValues()`?

Reply:
Yes please.